gokart/task.py

```diff
@@ -14,10 +15,33 @@
 from gokart.pandas_type_config import PandasTypeConfigMap
 from gokart.parameter import TaskInstanceParameter, ListTaskInstanceParameter
 from gokart.target import TargetOnKart
+from gokart.redis_lock import make_redis_params
 
 logger = getLogger(__name__)
 
 
+class RunWithLock:
+    def __init__(self, func):
+        self._func = func
+
+    def __call__(self, instance):
+        instance._lock_at_dump = False
+        output_list = luigi.task.flatten(instance.output())
+        return _run_with_lock(partial(self._func, self=instance), output_list)
+
+    def __get__(self, instance, owner_class):
+        return partial(self.__call__, instance)
+
+
+def _run_with_lock(func, output_list: list):
+    if len(output_list) == 0:
+        return func()
+
+    output = output_list.pop()
+    wrapped_func = output.wrap_with_lock(func)
+    return _run_with_lock(func=wrapped_func, output_list=output_list)
+
+
```

Comment:
It's better to separate the files.

Reply:
Yes, I will move `RunWithLock` to separate file.